flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBIncrementalRestoreOperation.java

```diff
@@ -0,0 +1,507 @@
+		throws BackendBuildingException {
+		// Register CF handlers
+		for (int i = 0; i < metaInfoSnapshots.size(); ++i) {
+			try {
+				getOrRegisterStateColumnFamilyHandle(
+					columnFamilyDescriptors.get(i),
+					columnFamilyHandles.get(i),
+					metaInfoSnapshots.get(i));
+			} catch (RocksDBException e) {
+				String errMsg = "Failed to register CF handle.";
+				LOG.error(errMsg, e);
+				throw new BackendBuildingException(errMsg, e);
+			}
+		}
+	}
+
+	public RocksDBIncrementalRestorePrepareResult prepareFiles(
+		KeyedStateHandle rawStateHandle,
+		Path restoreInstancePath) throws IOException {
+		IncrementalLocalKeyedStateHandle localKeyedStateHandle;
+		List<StateMetaInfoSnapshot> stateMetaInfoSnapshots;
+		try {
+			KeyedBackendSerializationProxy<K> serializationProxy;
+			if (rawStateHandle instanceof IncrementalKeyedStateHandle) {
+				IncrementalKeyedStateHandle restoreStateHandle = (IncrementalKeyedStateHandle) rawStateHandle;
+
+				// read state data.
+				try (RocksDBStateDownloader rocksDBStateDownloader =
+						new RocksDBStateDownloader(numberOfTransferringThreads)) {
+					rocksDBStateDownloader.transferAllStateDataToDirectory(
+						restoreStateHandle,
+						restoreInstancePath,
+						cancelStreamRegistry);
+				}
+
+				serializationProxy = readMetaData(restoreStateHandle.getMetaStateHandle());
+				stateMetaInfoSnapshots = serializationProxy.getStateMetaInfoSnapshots();
+				columnFamilyDescriptors = createAndRegisterColumnFamilyDescriptors(stateMetaInfoSnapshots, columnOptions);
+
+				// since we transferred all remote state to a local directory, we can use the same code as for
+				// local recovery.
+				localKeyedStateHandle = new IncrementalLocalKeyedStateHandle(
+					restoreStateHandle.getBackendIdentifier(),
+					restoreStateHandle.getCheckpointId(),
+					new DirectoryStateHandle(restoreInstancePath),
+					restoreStateHandle.getKeyGroupRange(),
+					restoreStateHandle.getMetaStateHandle(),
+					restoreStateHandle.getSharedState().keySet());
+			} else if (rawStateHandle instanceof IncrementalLocalKeyedStateHandle) {
+				// Recovery from local incremental state.
+				localKeyedStateHandle = (IncrementalLocalKeyedStateHandle) rawStateHandle;
+				serializationProxy = readMetaData(localKeyedStateHandle.getMetaDataState());
+				stateMetaInfoSnapshots = serializationProxy.getStateMetaInfoSnapshots();
+				columnFamilyDescriptors = createAndRegisterColumnFamilyDescriptors(stateMetaInfoSnapshots, columnOptions);
+			} else {
+				throw new IllegalStateException("Unexpected state handle type, " +
+					"expected " + IncrementalKeyedStateHandle.class + " or " + IncrementalLocalKeyedStateHandle.class +
+					", but found " + rawStateHandle.getClass());
+			}
+		} catch (Exception e) {
+			// clean up
+			FileSystem restoreFileSystem = restoreInstancePath.getFileSystem();
+			if (restoreFileSystem.exists(restoreInstancePath)) {
+				restoreFileSystem.delete(restoreInstancePath, true);
```

Comment:
Very good addition! I would suggest to consider error handling for all the exception that can happen when dealing with filesystem and delete here, because they can hide the original exception, and the log statement would not be reached as well. I think you can see better when changing this method from `throws IOException` to `throws BackendBuildingException`